### ISSUE TITLE
Discussion: Is parsing conceptually part of running, or not?

### ISSUE THREAD

**Issue by luketpeterson** [trueagi-io/hyperon-experimental#527]
Background: The Parser converts textual source code into atoms, which are then interpreted.  The interpreter runs within a runner, which has a runtime state.  The question at hand is whether the parser should be considered part of the run process, and thus be able to read and modify the runtime state or whether it is a pre-process that happens before running.

The reason it matters is because some tokens in the source text make reference to objects in the runtime state.  The simplest example being `&self`.

In the current implementation, the Tokenizer (invoked by the Parser) replaces the text `&self` with a grounded atom that refers to the running Space.  However, different modules have different interpretations of `&self`.  For now, we can work around this because different modules also have different Tokenizers, but this design choice has major conceptual implications for the MeTTa language.

If the answer is “Yes”, ie that parsing is part of running, then the parser should replace the token with a value derived from the actively running context.  Which means parsing needs to occur within an active runner, not as a pre-process.

If the answer is “No”, ie that parsing is a pre-process and MeTTa’s native program format is strictly Atom-based, then the parser does not have enough information to resolve the `&self` token all the way to a specific grounded Space atom at parse time.  Therefore it should turn`&self` into some abstract atom, or leave it as a symbol atom.  And then finally the interpreter would substitute the abstract atom or symbol with the specific grounded atom, possibly based on an equality expression from the running module's Space.

My personal feeling is that “No” makes more sense.  It is more similar to other languages where the running program can't reconfigure the parse logic as the program runs.  Most languages I know about require macros, etc. to be applied upstream, either in a pre-pass (like C) or from a dependency that loads in advance (like Rust proc macros), and I feel this will be less error-prone.  

This does, however, mean we need to rethink the `bind!` operator, perhaps to allow a grounded atom to be substituted for a symbol using an equality defined in the space, so `bind!` would not be needed.

**Comment by Necr0x0Der** [comment]
For me, what matters is the behavior of metta-scripts rather than whether parsing should or should not be conceptually part of running. i'm ok with an option when the needed behavior is achieved by another mechanism. We may want to get rid of special tokens at all, although parsing strings or numbers is difficult to avoid. And the idea of having a fixed parser, which will incorporate all such special cases, doesn't look appealing to me. If we do have some tokens, why not to allow defining them by modules and extensions? Consequently, as @vsbogd pointed out, the set of tokens can be different for different modules.  Maybe, we are overusing this mechanism of extending the set of tokens. I'm in favour of not using `bind!` if we have a better mechanism. [Unfortunately?] we don't support equalities for symbol atoms (e.g. `(= &one 1)`), but we could still have, say, `(= (self) %unrepresentable-grounded-atom-space%)` put into the space of the module, which equates `(self)` and the corresponding grounded atom instead of having `&self` token, which is replaced by the same grounded atom during parsing. OTOH, what is the benefit of having `(= (self) ...)` inserted into the space? I'm not sure. We insert something, which cannot be directly represented in metta scripts. A possibility to introduce direct syntactic representation of grounded atoms like `&self` looks natural. Once again, I'm not advocating for using `bind!`. Recently, I prefer to place equalities in the space instead, e.g. `! (let $space (new-space) (add-atom &self (= (my-space) $space))` (which could be made more convenient, though) rather than `!(bind! &my-space (new-space))`. I'd prefer also a version of `import!`, which returns a space rather than binds it to a new token. But even if we get rid of token bindings in `import!` and reduce the use of `bind!`, because it is kind of misused due to absence of a convenient token-less alternative, the possibility to introduce new tokens by models and extensions is very handy. We cannot avoid this by inserting equalities or other runtime binding mechanisms, if we want to be able to extend the language by families of tokens like numbers, strings, etc. For example, we wanted to introduce regex atoms. We need a token family for this like `r"[^"]*"` or such. Of course, we can have constructors, which convert strings to the needed object, e.g. `(regex "^Hello[[\.|!]]?$")` instead of `r"^Hello[[\.|!]]?$"`. It has its own inconveniences, e.g. `(= (intent (regex "^Hello[[\.|!]]?$")) Hello)` will not work for `(intent "Hello")`, because `(regex "^Hello[[\.|!]]?$")` will not be executed by the pattern matcher. It can be overcome by different means, but it does not work for now.

Basically, introduction of new tokens can be replaced by other means. Sometimes it should be replaced, indeed. In other cases, it is somewhat questionable if it is worthy. Answering the question "Is parsing conceptually part of running, or not?" I'd say that "Yes" answer is more amusing, and I'd stay with "Yes" unless there are indeed deep technical troubles implied by this choice. I'd prefer to reduce the use of custom tokens and encourage a token-less approach, when possible. But I wouldn't forcefully remove runtime tokenization. I understand that the question may arise from the work on modules and imports, and it could be an important choice right now. But if this is really the case, we may need discussing what are the problems with runtime parsing for the module system.
